Update random pod scaledown KEP for beta

damemi opened the pull request:
Ref: https://github.com/kubernetes/enhancements/issues/2185

damemi reviewed (commented):
/cc @alculquicondor @soltysh 
I've tried updating this with the PRR answers required for beta. Though, imo many of them seem irrelevant since this is a minor change to a previously-undocumented behavior. For example, I don't think we need a metric to expose this behavior (but it could be easy to add a simple boolean, or in some way measure the logarithmic performance of cumulative scaledowns on the cluster). Please let me know if you feel there are better responses we could provide.

alculquicondor commented:
You also need to update the file in `/keps/prod-readiness`

alculquicondor commented on keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md:
```diff
@@ -263,37 +263,36 @@ _This section must be completed when targeting alpha to a release._
-  Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
-  checking if there are objects with field X set) may be a last resort. Avoid
-  logs or events for this purpose.
+  Only by the presence of the feature gate
```
Explicitly say that all replicasets are affected.

alculquicondor commented on keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md:
```diff
@@ -263,37 +263,36 @@ _This section must be completed when targeting alpha to a release._
+    - Details: The health of this feature is dependent on the overall health of the 
+  control plane services, specifically kube-controller-manager but also to an extent 
+      services like API server, kubelet, etc.
+      We could add a metric to KCM to indicate that this feature gate is enabled
```
Remove this speculation. I don't think we should do that.

alculquicondor commented on keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md:
```diff
@@ -263,37 +263,36 @@ _This section must be completed when targeting alpha to a release._
+  Rollouts should have no effect, as this only relates to scaledown of replicasets
 
 * **What specific metrics should inform a rollback?**
+N/A
```
If there were metrics for number of pods deleted, maybe those could be used. Like, if there's a problem on the implementation, potentially you could see an uptick of pod deletions.

Another thing could be that the availability of applications is affected. Although the intent of the feature is that availability increases (because of better spreading).

wojtek-t commented:
/assign

wojtek-t commented on keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md:
```diff
@@ -263,37 +263,36 @@ _This section must be completed when targeting alpha to a release._
-  - [ ] Other (treat as last resort)
-    - Details:
+  - [x] Other (treat as last resort)
+    - Details: The health of this feature is dependent on the overall health of the 
```
This isn't helpful - it's true for almost any feature.

What we need is a signal to determine if the kcm is doing what it should be doing.

wojtek-t replied:
Can we come up with some metrics for this? Even if whitebox-y, that's still much better than nothing.

Maybe we can expose histogram of something like:
- |age of deleted pod| / |age of yougest pod|
[We expect all values to be within [1,2], so having values outside of it is a good signal that something is wrong.

[Well - technically it will collide with https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2255-pod-cost so maybe we additionally need a metric label whether such annotation was set or not]
[And then the SLO would be that there are no values >2 when the pod_cost_set=false]

damemi replied:
> |age of deleted pod| / |age of yougest pod|

are these currently available, or is there work needed to add them to KCM? @soltysh where are the KCM metrics set up so that we can start work on that?

alculquicondor commented on keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md:
```diff
@@ -263,37 +263,36 @@ _This section must be completed when targeting alpha to a release._
-  Describe manual testing that was done and the outcomes.
-  Longer term, we may want to require automated upgrade/rollback tests, but we
-  are missing a bunch of machinery and tooling and can't do that now.
+  N/A
```
Sure, I can do it once the code changes are in.

damemi replied:
> Sure, I can do it once the code changes are in.

Would that just be the beta promotion change https://github.com/kubernetes/kubernetes/pull/101767? I'm also confused about how this should be tested with the upgrade->downgrade->upgrade path

wojtek-t replied:
Please add a text that you will test that before the feature will graduate to beta.
Also - please update the KEP within the release cycle [see https://github.com/kubernetes/enhancements/pull/2538 as an example where we did that.]

wojtek-t commented on keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md:
```diff
@@ -263,37 +263,36 @@ _This section must be completed when targeting alpha to a release._
+      services like API server, kubelet, etc.
+      We could add a metric to KCM to indicate that this feature gate is enabled
 
 * **What are the reasonable SLOs (Service Level Objectives) for the above SLIs?**
```
I understand the intention, but I don't think it's a good SLO - SLI/SLO needs precise definition and this value will heavily depend on the load (and can change over time).

That may be a good metric informing rollback though...

wojtek-t commented on keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md:
```diff
@@ -263,37 +263,39 @@ _This section must be completed when targeting alpha to a release._
-  checking if there are objects with field X set) may be a last resort. Avoid
-  logs or events for this purpose.
+  The scaledown behavior of all replicasets will be affected by this featuregate being 
+  enabled, so somehow monitoring them will be necessary to determine it
```
I think the existing answer is fine - it will be used by all workloads.

Now we just need a way to verify that it really works - but it's more the SLI question below.

wojtek-t commented on keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md:
```diff
@@ -263,46 +263,40 @@ _This section must be completed when targeting alpha to a release._
-  - [ ] Metrics
-    - Metric name:
+  - [x] Metrics
+    - Metric name: deleted_pod_age_ratio
```
If this is going to be the metric that I proposed in: https://github.com/kubernetes/enhancements/pull/2691#discussion_r629877889

can you please define it somewhere in the KEP (and probably add a beta criteria that this metric is implemented)?

[If it's going to be a different metric, please define that too :) ]

wojtek-t commented on keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md:
```diff
@@ -263,46 +263,40 @@ _This section must be completed when targeting alpha to a release._
-  - 99% percentile over day of absolute value from (job creation time minus expected
-    job creation time) for cron job <= 10%
-  - 99,9% of /health requests per day finish with 200 code
+  There should be no values `>2` in the above metric when the Pod Cost annotation is set 
```
s/set/unset/, right?

[Or to be more specific - if this is unset for all pods of the replicaset...]

Basically (it's a comment toward my ask above to define the metric in the kep), I think that the metric should have a field of "pod-cost-annotation-exists" (better name welcome :-) ) that replicaset controller will be computing and setting when exporing  that metric (and then the check is trivial).

wojtek-t reviewed (commented):
Some suggestions/answers - hope these are useful.

soltysh commented on keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md:
```diff
@@ -263,37 +263,39 @@ _This section must be completed when targeting alpha to a release._
 * **How can a rollout fail? Can it impact already running workloads?**
-  Try to be as paranoid as possible - e.g., what if some components will restart
-   mid-rollout?
+  
```
Maybe something along:

It shouldn't impact already running workloads. One possible problem during rollout is a panic in the algorithm which might crash controller-manager.

soltysh reviewed (approved):
One suggestion for that one missing bit, the rest
/lgtm
/approve

damemi commented:
Updated with the latest feedback, and squashed.

One note about the metric (mentioned this to @alculquicondor offline) is that the defined range of `[1,2]` is only valid for pods where their deletion was based on sorting by timestamp. 

Similar to how this conflicts with the Pod Cost Annotation, there are [other criteria](https://github.com/damemi/kubernetes/blob/a8d105ab724ccdb45b4ff380d84fd356e833991e/pkg/controller/controller_utils.go#L792-L801) which could put a pod older than `2 * youngest_pod_age` as higher priority for deletion, such as a `Pending` status phase. I've tried to make this clear in the definitions of the new metric.

soltysh commented:
/lgtm

wojtek-t commented:
/lgtm
/approve

Thanks!
